### DISCUSSION TITLE
Transmit data back from a webApp to the telegram bot

### DISCUSSION THREAD

**Discussion by TheTinkerBasement** [telegraf/telegraf#1776] [Q&A]
Hi, 
I try to implement a telegram webApp which can send data back to the bot, see here https://core.telegram.org/bots/webapps#keyboard-button-web-apps

My bot uses a demo telegram webapp (running on github pages) which has a "send time" button function, see 
https://github.com/revenkroz/telegram-web-app-bot-example/blob/685d312c1bee1537ebd226ab8a5e5fcc1451b565/index.html#L270 
here is the webapp URL: https://revenkroz.github.io/telegram-web-app-bot-example/index.html
from the `telegram-web-app.js`  javascript library running in the browser.

Problem is that my `inline_query` event handler in the bot never fires, when I click the "send time" button in my webapp.

here is the code to test for yourself (should work out of the box, just add your API-KEY, then run the bot and hit command `/register`)

```javascript
const { Telegraf, Markup } = require("telegraf");

const bot = new Telegraf("API-KEY");

bot.command("register", (ctx) => {
  return ctx.reply(
    "open webapp",
    Markup.inlineKeyboard([
      Markup.button.webApp(
        "Open",
        "https://revenkroz.github.io/telegram-web-app-bot-example/index.html"
      ),
    ])
  );
});

bot.on("inline_query", async (ctx) => {
  // this never fires here
  console.log("hello from webapp");
  const result = [];
  // Explicit usage
  await ctx.telegram.answerInlineQuery(ctx.webAppData.id, result);
});

bot.launch();
```

**Comment by TheTinkerBasement** [comment]
I figured it, classical case of RTFM.
I tried "Keyboard Button Web Apps" from docs here: https://core.telegram.org/bots/webapps#keyboard-button-web-apps
this only works with keyboards and service messages so solution is:

```javascript
const { Telegraf, Markup } = require("telegraf");

const bot = new Telegraf("5820181457:AAGLYpxDD2MmUFrgPhrM17U94ExGcIAi9y0");

bot.command("register", (ctx) => {
  return ctx.reply(
    "open webapp",
    Markup.keyboard([
      Markup.button.webApp(
        "Open",
        "https://revenkroz.github.io/telegram-web-app-bot-example/index.html"
      ),
    ])
  );
});
bot.on("message", async (ctx) => {
  console.log(ctx.message.web_app_data)
  return ctx.reply(ctx.message.web_app_data.data)

});

bot.launch();
```

**Comment by TheTinkerBasement** [comment]
Now I want to do the same using "Inline Button Web Apps"

https://core.telegram.org/bots/webapps#inline-button-web-apps

`Quote: "The bot can call the Bot API method [answerWebAppQuery](https://core.telegram.org/bots/api#answerwebappquery) to send an inline message from the user back to the bot and close the Web App. After receiving the message, the bot can continue communicating with the user."`

How can I create a callback to get the results of a `answerWebAppQuery` from the webapp (https://core.telegram.org/bots/api#answerwebappquery) in my bot?

**Reply by MKRhere** [reply]
Only keyboard button web apps can use `sendData` (which results in a `web_app_data` message).

Other web apps must make an API call back to its own server. They should send `initDataUnsafe.query_id` to their server, using which server may call `answerWebAppQuery`.

**Reply by barinbritva** [reply]
Hi MKRhere, do you have any examples of such an implementation?

**Reply by MKRhere** [reply]
See our docs: https://github.com/feathers-studio/telegraf-docs/tree/673ad547190568183ce699e42dd3920c4804cbec/examples/mini-apps#feature-table

**Reply by barinbritva** [reply]
Thanks a lot!

**Comment by ldijkman** [comment]
is it normal that the webapp miniapp closes direct
after use Telegram.WebApp.sendData()
https://t.me/LuberthDijkmanbot/schedule
https://t.me/Luberth_Dijkman/94

**Reply by creativebash** [reply]
Yes. Once the mini app calls the sendData() method, the bot receives the data and proceeds to communicate with the user. This process closes the mini app.

"To transmit data from the user back to the bot, the Mini App can call the [Telegram.WebApp.sendData](https://core.telegram.org/bots/webapps#initializing-mini-apps) method. Data will be transmitted to the bot as a String in a service message. **The bot can continue communicating with the user after receiving it** "

The use cases therefore includes _a personalized calendar for selecting dates; selecting data from a list with advanced search options; a randomizer that lets the user “spin a wheel” and chooses one of the available options, etc._
